src/handlers/v1/markets/[chainId]/[marketId]/zap-withdraw/index.ts

```diff
@@ -0,0 +1,198 @@
+import { validateMarket } from 'src/validations';
+
+type GetZapWithdrawQuotationRouteParams = EventPathParameters<{ chainId: string; marketId: string }> &
+  EventBody<{
+    account?: string;
+    withdrawalToken?: common.TokenObject;
+    amount?: string;
+    targetToken?: common.TokenObject;
+    slippage?: number;
+  }>;
+
+type GetZapWithdrawQuotationResponseBody = QuoteAPIResponseBody<ZapQuotation>;
+
+export const v1GetZapWithdrawQuotationRoute: Route<GetZapWithdrawQuotationRouteParams> = {
+  method: 'POST',
+  path: '/v1/markets/{chainId}/{marketId}/zap-withdraw',
+  handler: async (event) => {
+    const chainId = Number(event.pathParameters.chainId);
+    const marketId = event.pathParameters.marketId.toUpperCase();
+    if (!validateMarket(chainId, marketId)) {
+      throw newHttpError(400, { code: '400.1', message: 'market does not exist' });
+    }
+
+    if (!event.body) {
+      throw newHttpError(400, { code: '400.2', message: 'body is invalid' });
+    }
+
+    let account = event.body.account;
+    if (!account) {
+      throw newHttpError(400, { code: '400.3', message: `account can't be blank` });
+    }
+    try {
+      account = utils.getAddress(account);
+    } catch {
+      throw newHttpError(400, { code: '400.4', message: 'account is invalid' });
+    }
+
+    const service = new Service(chainId);
+
+    let marketInfo: MarketInfo;
+    try {
+      marketInfo = await service.getMarketInfo(marketId, account);
+    } catch (err) {
+      throw newInternalServerError(err);
+    }
+    const { utilization, healthRate, liquidationThreshold, borrowUSD, collateralUSD, netAPR } = marketInfo;
+    const currentPosition = { utilization, healthRate, liquidationThreshold, borrowUSD, collateralUSD, netAPR };
+
+    let targetTokenAmount = '0';
+    const logics: GetZapWithdrawQuotationResponseBody['logics'] = [];
+    let approvals: GetZapWithdrawQuotationResponseBody['approvals'] = [];
+    let permitData: GetZapWithdrawQuotationResponseBody['permitData'];
+    let targetPosition = currentPosition;
+    if (event.body.withdrawalToken && event.body.amount && event.body.targetToken && Number(event.body.amount) > 0) {
+      const { amount, slippage } = event.body;
+      const {
+        baseToken,
+        supplyAPR,
+        supplyBalance,
+        supplyUSD,
+        borrowAPR,
+        borrowCapacityUSD,
+        liquidationLimit,
+        baseTokenPrice,
+        collaterals,
+      } = marketInfo;
+
+      const withdrawalToken = common.Token.from(event.body.withdrawalToken);
+      const targetToken = common.Token.from(event.body.targetToken);
+
+      // 1. check withdraw collateral or base token
+      const withdrawalCollateral = collaterals.find(({ asset }) => asset.is(withdrawalToken.unwrapped));
+      if (!withdrawalCollateral && !withdrawalToken.unwrapped.is(baseToken)) {
+        throw newHttpError(400, { code: '400.5', message: 'withdrawal token is not collateral nor base' });
+      }
```

Comment:
Why not leaving the throw to the condition check after L130?

Reply:
Since the logic between L91 and L130 is preparing the withdrawal logic, we need to make sure the `withdrawalToken` is collateral token or base token beforehand.

Reply:
L93 is `if (withdrawalToken.unwrapped.is(baseToken))`. We can change L113 `else` to `else if (withdrawalCollateral)`, then `else` at L119 to throw.

Reply:
I see. Yes, we can do that.

The advantage would be we don't need to check the token one more time at L87. 
The disadvantage would be the throw hides in the if-else clause. Also, the logic flow becomes a little bit different compared to other zaps.

I'm open to either way. What do @chouandy think?

Reply:
I usually prefer a positive approach and minimal operations when coding, so I would suggest the following:

1. First, check if it's the base token.
2. Then, check if it's the collateral token.
3. If neither, throw an error.

```typescript
if (is base token) {
   ...
} else {
  if (is collateral token) {
    ...
  } else {
    throw newHttpError(400, { code: '400.5', message: 'withdrawal token is not collateral nor base' });
  }
}
```

Reply:
OK. Will fix it in the next commit.